clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/view/MaterializedViewWriter.java

```diff
@@ -67,34 +67,40 @@ public CompletableFuture<Void> processRecord(
       int oldValueSchemaId,
       GenericRecord replicationMetadataRecord,
       Lazy<GenericRecord> valueProvider) {
-    return processRecord(newValue, key, newValueSchemaId, false, valueProvider);
+    return processRecord(newValue, key, newValueSchemaId, null, valueProvider);
   }
 
   /**
-   * Before we have proper chunking support for view writers we assume that even when chunking is enabled the actual
-   * k/v will not be chunked. This way we don't need to worry about how to ensure all the chunks are forwarded to the
-   * same view partition during NR's pass-through mode. Proper chunking support can leverage message footer populated
-   * by the CompositeVeniceWriter in VPJ (write to views first and then VT) to figure out which view partition to
-   * forward the chunks to. Another alternative is trigger the view writer write upon receiving the manifest, and we
-   * will assemble and re-chunk.
+   * During NR pass-through viewPartitionSet is going to be provided. This way we can forward record or chunks of a
+   * record to the appropriate view partition without the need to assemble or repartition.
    */
   @Override
   public CompletableFuture<Void> processRecord(
       ByteBuffer newValue,
       byte[] key,
       int newValueSchemaId,
-      boolean isChunkedKey,
+      Set<Integer> viewPartitionSet,
       Lazy<GenericRecord> newValueProvider) {
-    byte[] viewTopicKey = key;
-    if (isChunkedKey) {
-      viewTopicKey = keyWithChunkingSuffixSerializer.getKeyFromChunkedKey(key);
-    }
     byte[] newValueBytes = newValue == null ? null : ByteUtils.extractByteArray(newValue);
+    if (viewPartitionSet != null) {
+      if (newValue == null) {
+        // This is unexpected because we only attach view partition map for PUT records.
+        throw new VeniceException(
+            "Encountered a null PUT record while having view partition map in the message header");
+      }
+      // Forward the record to corresponding view partition without any processing (NR pass-through mode).
+      return veniceWriter.get().forwardPut(key, newValueBytes, newValueSchemaId, viewPartitionSet);
+    }
     if (newValue == null) {
       // this is a delete operation
-      return veniceWriter.get().complexDelete(viewTopicKey, newValueProvider);
+      return veniceWriter.get().complexDelete(key, newValueProvider);
```

Comment:
Nit,
maybe we should add more details for `delete` since `newValueProvider` is actually providing the previous value.